ams/server/src/main/java/com/netease/arctic/server/manager/ActivePluginManager.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netease.arctic.server.manager;
+
+import com.netease.arctic.ams.api.ActivePlugin;
+import com.netease.arctic.ams.api.PluginManager;
+import com.netease.arctic.server.exception.LoadingPluginException;
+import com.netease.arctic.server.utils.PreconditionUtils;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public abstract class ActivePluginManager<T extends ActivePlugin>
+    implements PluginManager<T>, Iterable<T> {
+
+  protected static final String PUGIN_IMPLEMENTION_CLASS = "impl";
+
+  private final Map<String, T> installedPlugins = new ConcurrentHashMap<>();
+
+  protected ActivePluginManager() {
+  }
+
+  protected abstract Map<String, String> loadProperties(String pluginName);
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public void install(String pluginName) {
+    PreconditionUtils.checkNotExist(installedPlugins.containsKey(pluginName),
+        "Plugin " + pluginName);
+    Map<String, String> properties = loadProperties(pluginName);
+    String pluginClass = properties.get(PUGIN_IMPLEMENTION_CLASS);
+    try {
+      Class<?> clazz = Class.forName(pluginClass);
+      T plugin = (T) clazz.newInstance();
+      installedPlugins.computeIfAbsent(pluginName, k -> {
```

Comment:
The code earlier has already performed a checkNotExist, is the use of computeIfAbsent at this point for multi-threaded scenarios? If it is for multi-threading, can we consider putting the class loading and other into the same function?

Reply:
yeah, that will be better